### PR TITLE
Add S3 nodes to AWS Explorer tree

### PR DESCRIPTION
**Pull request by ckant**
#### S3 Nodes (per region)
- Inline: `"Create Bucket"`
- Context Menu: `"Create Bucket"`

#### Bucket Nodes
- Inline: `"Create Folder"`, `"Upload File"`
- Context Menu: `"Upload File"`, `"Create Folder"` | `"Copy Name"`, `"Copy Arn"`

#### Folder Nodes
- Inline: `"Create Folder"`, `"Upload File"`
- Context Menu: `"Upload File"`, `"Create Folder"` | `"Copy Name"`, `"Copy Path"`, `"Copy Arn"`

#### File Nodes
- Inline: `"Download File As"`
- Context Menu: `"Download File As"` | `"Copy Name"`, `"Copy Path"`, `"Copy Arn"`

#### More Results Nodes
- Command: `"Load More"`

---

Bucket and Folder Nodes cache their children and support pagination.

The very first call to `getChildren()` always loads and returns the initial page of children.
Subsequent calls to `getChildren()` always load from the cache.

A call to `loadMoreChildren()` appends the next page of children to the cache.
The node must subsequently be refreshed to sync the view with the cache.

A full refresh on a Bucket or Folder Node is done by a call to `clearCache()`.
The nodes then refresh their state and behave as if they had just been created.

## Related Issue(s)

* [Create reusable fakes/stubs for VSCode UI components](https://github.com/aws/aws-toolkit-vscode/pull/1074) #1074
* [Create S3 client wrapper](https://github.com/aws/aws-toolkit-vscode/pull/1075) #1075
* [Introduce commands for acting on S3 nodes / resources ](https://github.com/aws/aws-toolkit-vscode/pull/1080) #1080

## Screenshots (if appropriate)

s3demolarge

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.nls.json [hunk 1]
```diff
@@ -142,9 +148,18 @@
     "AWS.explorerNode.cloudFormation.error": "Error loading CloudFormation resources",
     "AWS.explorerNode.container.noItems": "[no items]",
     "AWS.explorerNode.copiedArn": "Copied ARN to clipboard",
+    "AWS.explorerNode.copiedName": "Copied name to clipboard",
+    "AWS.explorerNode.copiedPath": "Copied path to clipboard",
```
Comment (bryceitoc9) [inline package.nls.json:152]: Nit: make this generic: `"Copied {0} to clipboard"`

File: package.nls.json [hunk 2]
```diff
@@ -251,6 +266,26 @@
...
     "AWS.profile.recentlyUsed": "recently used",
     "AWS.running.command": "Running command: {0}",
+    "AWS.s3.createBucket.placeHolder": "Bucket Name",
+    "AWS.s3.createBucket.prompt": "Create Bucket",
+    "AWS.s3.createBucket.error.illegalCharacters": "Bucket name must only contain lowercase letters, numbers, hyphens, and periods",
+    "AWS.s3.createBucket.error.invalidEnd": "Bucket name must end with a lowercase letter or number",
+    "AWS.s3.createBucket.error.invalidLength": "Bucket name must be between 3 and 63 characters long",
+    "AWS.s3.createBucket.error.invalidStart": "Bucket name must start with a lowercase letter or number",
+    "AWS.s3.createBucket.error.misusedPeriods": "Periods in bucket name must be surrounded by a lowercase letter or number",
+    "AWS.s3.createBucket.error.resemblesIpAddress": "Bucket name must not resemble an IP address",
```
Comment (bryceitoc9) [inline package.nls.json:276]: I feel like we already have these lines of text for the deploy workflow? If so, no point in duplicating them

File: src/extension.ts [hunk 3]
```diff
@@ -220,6 +220,15 @@ function initializeIconPaths(context: vscode.ExtensionContext) {
...
     ext.iconPaths.light.registry = context.asAbsolutePath('resources/light/registry.svg')
 
+    ext.iconPaths.dark.s3.bucket = context.asAbsolutePath('resources/dark/s3/bucket.svg')
+    ext.iconPaths.light.s3.bucket = context.asAbsolutePath('resources/light/s3/bucket.svg')
+
+    ext.iconPaths.dark.s3.folder = context.asAbsolutePath('third-party/resources/from-vscode/dark/folder.svg')
+    ext.iconPaths.light.s3.folder = context.asAbsolutePath('third-party/resources/from-vscode/light/folder.svg')
+
+    ext.iconPaths.dark.s3.file = context.asAbsolutePath('third-party/resources/from-vscode/dark/document.svg')
+    ext.iconPaths.light.s3.file = context.asAbsolutePath('third-party/resources/from-vscode/light/document.svg')
```
Comment (bryceitoc9) [inline src/extension.ts:230]: Now that we're pinned to VS Code v1.42.0, could we try codicons instead of importing new SVGs?

https://code.visualstudio.com/updates/v1_42#_codicons-for-reuse

File: src/s3/explorer/s3FileNode.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        public readonly bucket: Bucket,
+        public readonly file: File,
+        private readonly s3: S3Client,
+        now: Date = new Date()
+    ) {
+        super(file.name)
+        if (file.sizeBytes !== undefined && file.lastModified) {
+            const readableSize = formatBytes(file.sizeBytes)
+
+            // Prevent clock skew showing future date
```
Comment (bryceitoc9) [inline src/s3/explorer/s3FileNode.ts:29]: Why would this show a future date?

File: src/s3/explorer/s3BucketNode.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    public async loadMoreChildren(): Promise<AWSTreeNodeBase[]> {
+        const response = await this.s3.listObjects({
+            bucketName: this.bucket.name,
+            continuationToken: this.cache.continuationToken,
+        })
+
+        const newFolders = response.folders.map(folder => new S3FolderNode(this.bucket, folder, this.s3))
+        const newFiles = response.files.map(file => new S3FileNode(this.bucket, file, this.s3))
+        this.cache.appendItems(newFolders, newFiles, response.continuationToken)
```
Comment (bryceitoc9) [inline src/s3/explorer/s3BucketNode.ts:81]: Nit: Move this logic to the cache (just send the response object without filtering) since this is duplicated in the folder node. (This didn't get added in the last review for some reason...)

File: package.json [hunk 6]
```diff
@@ -578,10 +598,40 @@
                     "group": "0@2"
                 },
                 {
-                    "command": "aws.copyArn",
-                    "when": "view == aws.explorer && viewItem =~ /^(awsRegionFunctionNode|awsCloudFormationFunctionNode|awsStateMachineNode|awsCloudFormationNode)$/",
+                    "command": "aws.s3.downloadFileAs",
+                    "when": "view == aws.explorer && viewItem == awsS3FileNode",
+                    "group": "0@1"
```
Comment (awschristou) [inline package.json:603]: I'm not sure what these context menus look like, but similar commands should be grouped together between separators. Something like   (download, upload) | (create bucket, folder) | (copy X)

File: package.json [hunk 7]
```diff
@@ -645,6 +695,47 @@
...
+                "icon": {
+                    "light": "third-party/resources/from-vscode-icons/light/new-folder.svg",
+                    "dark": "third-party/resources/from-vscode-icons/dark/new-folder.svg"
+                }
+            },
+            {
+                "command": "aws.s3.createBucket",
+                "title": "%AWS.command.s3.createBucket%",
+                "category": "AWS",
+                "icon": {
```
Comment (awschristou) [inline package.json:734]: Could there be a "bucket with a plus symbol" to better represent "Create bucket" ?

File: package.json (hunk 6 above)
Comment (ckant) [reply package.json:603]: I tried to put the groups in the description. Thanks for reminding me, was going to put create bucket / folder into its own section. Right now I have it grouped with upload.

File: package.json (hunk 7 above)
Comment (ckant) [reply package.json:734]: Yes, that's one of the things I wanted to get a new icon for. What do you think about the download and upload icons (cloud with up or down arrow)? I kind of think that should be the doc with an up or down arrow. 

I'm noticing now that GitHub actually uses a similar cloud with an up arrow for uploading files to this comment box haha.

File: src/s3/explorer/s3FileNode.ts (hunk 4 above)
Comment (ckant) [reply src/s3/explorer/s3FileNode.ts:29]: Small variations in server time versus machine time. I observed that as soon as I uploaded something, it would show last updated date as "a few seconds from now". 

I'm sure there are other edge cases where it will show a future date, but this should hopefully squash the most common/annoying one.

Another option would be to clamp the time to be <= now, but I think that would be even more confusing if it was legitimately hitting an edge case. Time is hard.

**Comment by ckant** [comment]
> Looks good! If you have a sec, maybe add S3Error cases to the tests?

This is a good point, but I think the action would be to document that these methods throw `S3Error` rather than test for that. Since the nodes propagate all exceptions, there's no translation to test for. Callers, however, should be made aware of what to expect (which is identical as what to expect from `S3Client`).

File: package.nls.json (hunk 2 above)
Comment (ckant) [reply package.nls.json:276]: Will consolidate.

File: src/s3/explorer/s3BucketNode.ts (hunk 5 above)
Comment (ckant) [reply src/s3/explorer/s3BucketNode.ts:81]: I think there's a more dramatic de-duplication that should be done between Folder and Bucket considering how similar they are to one another. I had tried to add stuff to an abstract class, but it was creating a circular dependency so I had scrapped it. This is one of the areas I wasn't satisfied with. Will look again.

File: src/extension.ts (hunk 3 above)
Comment (ckant) [reply src/extension.ts:230]: Good idea. Matches the icon set.

image

File: src/s3/explorer/s3BucketNode.ts (hunk 5 above)
Comment (ckant) [reply src/s3/explorer/s3BucketNode.ts:81]: I kept `Bucket` and `File` as separate *(similar)* nodes. However, I extracted out the caching and "load more..." functionality which was what I was worried about.

File: package.json (hunk 7 above)
Comment (ckant) [reply package.json:734]: Added the create bucket icon and kept the cloud icons for now

File: src/s3/explorer/s3BucketNode.ts [hunk 8]
```diff
@@ -18,82 +19,85 @@ import { ErrorNode } from '../../shared/treeview/nodes/errorNode'
...
-    private async getExistingChildren(): Promise<AWSTreeNodeBase[]> {
-        return this.cache.nodes
+    public clearChildren(): void {
+        this.childLoader.clearChildren()
     }
 
-    public async loadMoreChildren(): Promise<AWSTreeNodeBase[]> {
+    private async loadPage(continuationToken: string | undefined): Promise<ChildNodePage> {
+        getLogger().debug(`Loading page for %O using continuationToken %s`, this, continuationToken)
         const response = await this.s3.listObjects({
```
Comment (bryceitoc9) [inline src/s3/explorer/s3BucketNode.ts:71]: Just making sure I'm clear, what's the error handling strategy?

File: src/s3/explorer/s3BucketNode.ts [hunk 9]
```diff
@@ -107,7 +111,11 @@ export class S3BucketNode extends AWSTreeNodeBase implements AWSResourceNode, Lo
         return ''
     }
 
-    public toString(): string {
+    public [inspect.custom](): string {
```
Comment (bryceitoc9) [inline src/s3/explorer/s3BucketNode.ts:114]: Never seen this syntax before, what's this do?

File: package.json [hunk 10]
```diff
@@ -612,27 +624,22 @@
...
                 {
                     "command": "aws.copyArn",
                     "when": "view == aws.explorer && viewItem =~ /^(awsRegionFunctionNode|awsCloudFormationFunctionNode|awsStateMachineNode|awsCloudFormationNode|awsS3BucketNode|awsS3FolderNode|awsS3FileNode)$/",
-                    "group": "1@3"
-                },
-                {
-                    "command": "aws.s3.createBucket",
-                    "when": "view == aws.explorer && viewItem == awsS3Node",
-                    "group": "1@4"
+                    "group": "2@3"
```
[the remark this answers is not in the document]
Comment (ckant) [reply package.json:642]: Can you clarify the suggestion? This is what it looks like. I think it's reasonable since all the "create-like" stuff goes at the top and all the copy stuff is just below it. Perhaps we can revisit at a later time

image

[the remark this answers is not in the document]
Comment (bryceitoc9) [reply package.json:642]: It's a nit, if everything else has:
Copy Name
Copy ARN
in that order, I'd recommend keeping it that way and moving Copy Path (either underneath instead of between the two or to a different section, probably underneath if anything)

File: package.json [hunk 11]
```diff
@@ -1021,6 +1019,7 @@
         "cloudformation-schema-js-yaml": "^1.0.1",
         "cross-spawn": "^6.0.5",
         "del": "^3.0.0",
+        "downloads-folder": "^1.0.4",
```
[the remark this answers is not in the document]
Comment (ckant) [reply package.json:1022]: Looks like the user can customize the download directory in *nix and windows. Might not be common, but seems useful.

File: package.nls.json [hunk 12]
```diff
@@ -17,6 +17,7 @@
     "AWS.configuration.description.logLevel": "The AWS Toolkit's log level (changes reflected on restart)",
     "AWS.configuration.description.onDefaultRegionMissing": "Action to take when a Profile's default region is hidden in the Explorer. Possible values:\n* `add` - shows region in the explorer\n* `ignore` - does nothing with the region\n* `prompt` - (default) asks the user what they would like to do.",
     "AWS.configuration.sam.template.depth": "The maximum subfolder depth within a workspace that the Toolkit will search for SAM Template files",
+    "AWS.configuration.description.s3.maxItemsPerPage": "Controls how many S3 items are listed before showing a node to `Load More...`.\nThis corresponds to the `MaxKeys` requested in a single call to S3. [Learn More](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#AmazonS3-ListObjectsV2-response-MaxKeys)",
```
[the remark this answers is not in the document]
Comment (ckant) [reply package.nls.json:20]: Fair, but doesn't seem to hurt. I like the added transparency.

File: package.nls.json [hunk 13]
```diff
@@ -150,18 +151,14 @@
     "AWS.explorerNode.cloudWatchLogs.noGroups": "[No Log Groups found]",
     "AWS.explorerNode.cloudWatchLogs.error": "Error loading CloudWatch Logs resources",
     "AWS.explorerNode.container.noItems": "[no items]",
-    "AWS.explorerNode.copiedArn": "Copied ARN to clipboard",
-    "AWS.explorerNode.copiedName": "Copied name to clipboard",
-    "AWS.explorerNode.copiedPath": "Copied path to clipboard",
+    "AWS.explorerNode.copiedToClipboard": "$(clippy) Copied {0} to clipboard",
```
[the remark this answers is not in the document]
Comment (ckant) [reply package.nls.json:154]: The fact that they called it clippy made me question whether it was going to be renamed in the future lol

I've added it to all of those commands. They all share this string.

File: resources/dark/s3/bucket.svg [hunk 14]
```diff
@@ -4,8 +4,8 @@
     <title>Amazon-S3-Standard-Icon_16_Squid</title>
     <desc>Created with Sketch.</desc>
     <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
-        <g id="AWS-Storage-Product-Icons_Squid" transform="translate(-44.000000, -120.000000)" fill="#999999">
```
[the remark this answers is not in the document]
Comment (ckant) [reply resources/dark/s3/bucket.svg]: Matches the VSCode color scheme for icons. It's not immediately apparent on the dark color scheme.

File: resources/light/s3/create-bucket.svg [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+        <polygon id="Path" fill="#424242" points="4.13501792 7.08891292 3.13501792 7.08891292 3.13501792 4.08891292 0.135017917 4.08891292 0.135017917 3.08891292 3.13501792 3.08891292 3.13501792 0.0889129225 4.13501792 0.0889129225 4.13501792 3.08891292 7.13501792 3.08891292 7.13501792 4.08891292 4.13501792 4.08891292"></polygon>
```
[the remark this answers is not in the document]
Comment (ckant) [reply resources/light/s3/create-bucket.svg:13]: Matches the VSCode color scheme. 
It's quite noticeable on the light color scheme.

The blue drowns out the gray.
image

versus

image

I will update the other icons to match this in a future PR since this bothers me 😄 .

File: src/awsexplorer/childNodeCache.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    private _isPristine: boolean = true
+
+    /**
+     * Appends a new page to the cache.
+     *
+     * Once this has been called, the cache is no longer considered pristine ({@link isPristine} will return false).
+     * This is true even if the original state of the cache remains unchanged (all items appended are empty/undefined).
+     */
+    public appendPage(page: ChildNodePage): void {
+        this._children = [...this._children, ...page.newChildren]
```
[the remark this answers is not in the document]
Comment (ckant) [reply src/awsexplorer/childNodeCache.ts:26]: Immutability of the object itself is usually preferable to immutability of the reference. The benefit is that the object won't change while you're looking at it. The reference might be reassigned, but no one else will be changing the object you're looking at.

(I realize _children is neither immutable nor a ReadonlyArray, but it's been treated as such.)

File: package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (ckant) [reply package.json:642]: Oh, sure. I think I put that because I saw ARN as the least valuable ID to copy (since S3 makes little use of them).

But there's merit in keeping the ordering consistent with other nodes.

File: src/awsexplorer/childNodeCache.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (ckant) [reply src/awsexplorer/childNodeCache.ts:26]: Another hidden benefit is that if you wrap the (change array) logic into a function, the changes become a black box. It's harder for the caller to reason about what changes have been made to the array.

If you wrap the (return array) logic into a function, there are no changes. The caller can choose how to deal with the result by filtering, merging with another array, etc.

Definitely not cut and dried though. The ideal is immutable reference and immutable object.

File: src/s3/explorer/s3BucketNode.ts (hunk 8 above)
Comment (ckant) [reply src/s3/explorer/s3BucketNode.ts:71]: Good callout. The initial page failure inserts an error node. But if a subsequent load more fails, it looks like the button doesn't do anything. 

`loadMoreChildren` failures probably need to trigger an error message to let the user know. Doesn't seem appropriate to put the error node in there.

File: src/s3/explorer/s3BucketNode.ts (hunk 9 above)
Comment (ckant) [reply src/s3/explorer/s3BucketNode.ts:114]: Overrides the output of `log.info('%O', node)` (See #1108)

Makes it look nicer in the logs. Without this, `%O` would show the full object similar to `JSON.stringify`. Similar to adding a dedicated field called `description` or something except you don't need to explicitly log that field.

Comment (ckant) [reply src/s3/explorer/s3BucketNode.ts:114]: Oh and the syntax is to define the method name with a name that is a Symbol.
https://javascript.info/symbol

Basically a way to declare you're implementing this interface without ambiguity (accidentally creating a method with the same name).

Comment (kiiadi) [reply src/s3/explorer/s3BucketNode.ts:114]: TIL - they're almost the opposite of a Scala [Symbol](https://www.scala-lang.org/api/current/scala/Symbol.html)

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (ckant) [reply package.json]: Useful, but annoying. This needed access to `registry-js` which is a native npm module for accessing the win registry. It introduced additional dependencies on the build and caused failures. Not worth the hassle. 

Removed this in favor of the simpler suggestion you made.
